### DISCUSSION TITLE
File System API helpers

### DISCUSSION THREAD

**Discussion by magne4000** [vercel/vercel#7573] [Help]
I'm working on a [Vercel integration for `vite-plugin-ssr`](https://github.com/brillout/vite-plugin-ssr/issues/222), and this takes advantage of the FS API.

I see there are a lot of tools around for building and bundling code (`@vercel/node`, `@vercel/build-utils`, `@vercel/routing-utils` to name a few), but most are not documented.
I'm wondering what tools I can use (and how) to help me with the following:
- Compile `/api/**` files to `.output/server/pages` and `.output/server/pages/api`. Currently doing that with `esbuild`.
- ~~Merge [user provided config](https://vercel.com/docs/cli#project-configuration) + [plugin provided config](https://github.com/magne4000/test-vite-vercel-plugin/blob/main/vite.config.ts). I would like to be able to have the user provide the config in the same format as defined in the official doc, but I'm missing the toolchain that generates all `.output/*-manifest.json` files.~~
  - actually not necessary since `vercel.json` still works as intended even with manually generated `.ouput` directory, I wrongly assumed the opposite.

**Comment by brillout** [comment]
@barelyhuman We are building an ISR Vercel deployer for Vite projects (https://github.com/brillout/vite-plugin-ssr/issues/222); maybe you have some insights about the questions @magne4000 is ask himself. Thanks for your help back then, it helped a lot.

**Reply by barelyhuman** [reply]
Thanks for tag, 
I need to understand a few things here , because otherwise my answer would not even be for the question you are asking. 

`.output` => direct mapping of what vercel considers as deployable
`*.manifest.json` => files generated by vercel / next to reference as to what points to what 

Am I on the right points? @brillout @magne4000

**Reply by brillout** [reply]
Thanks for the reply. @magne4000 will answer you.

**Reply by magne4000** [reply]
> Am I on the right points?

yes that's right

**Reply by barelyhuman** [reply]
Cool, 

So the manifests are all built using webpack plugins that are written by vercel and these are available in the build tools that are a part of the source code in at least next.js 

Here's a few of them that you might be looking for 

https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/plugins/build-manifest-plugin.ts
https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/plugins/flight-manifest-plugin.ts
https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/plugins/functions-manifest-plugin.ts
https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack/plugins/pages-manifest-plugin.ts

I don't remember any of them making it into a public package but you can basically pick up whatever code you need from them 

> Merge [user provided config](https://vercel.com/docs/cli#project-configuration) + [plugin provided config](https://github.com/magne4000/test-vite-vercel-plugin/blob/main/vite.config.ts). I would like to be able to have the user provide the config in...

For this you'll have to build a small subset of a config reader, I don't think a tool exists for vite for this specific usecase  ... but your best bet is to look for something from https://github.com/antfu or it's not that hard to write one if absolutely necessary


> @vercel/build-utils, @vercel/routing-utils

These are specifically for the vercel platform based runtimes and not really for app generation though you can use these to configure how builds are triggered, handled, etc

**Reply by magne4000** [reply]
Nice! I'll take a look at all of that, thanks

**Reply by magne4000** [reply]
After some tests I realize I wrongly assumed that `vercel.json` was not taken into account when manually generating `.ouput` directory. Now that I know that it still works as intended, I do not need to merge user config with whatever config the plugin generates

**Comment by magne4000** [comment]
I'm also having trouble with `routes-manifest.json` containing a fallback route.

#### routes-manifest.json

```json
{
  "version": 3,
  "basePath": "/",
  "pages404": true,
  "rewrites": [
    {
      "source": "/dynamic",
      "destination": "/api/ssr_",
      "regex": "^/dynamic.*$"
    },
    {
      "source": "/",
      "destination": "/api/ssr_",
      "regex": "^/((?!assets/)(?!api/).*)$"
    }
  ]
}
```


The second rewrite rule is the fallback rule. It should catch everything that is not `/api/...` or `/assets/...`. I assume it also should not catch `/dynamic` urls as the rule is defined before it. Currently, this second rule does nothing it seems. Is this intended behavior? It's not clear how overlapping rules behave (even more nebulous if they overlap between `rewrites` and `dynamicRoutes`)

I also tried to put the same rule in `vercel.json`, it supersedes everything defined in `routes-manifest.json`.

**Reply by barelyhuman** [reply]
<img width="394" alt="Screenshot 2022-03-21 at 10 48 33 PM" src="https://user-images.githubusercontent.com/43572006/159328417-1f72f76e-2fde-4929-8fdd-4f0aa11d1232.png">

Well the regex actually catches the `asset` and `api` routes , so you might wanna change that 

also the rewrites are based on the order they are written in , you can refer to the lines from here regrading the logic
https://github.com/vercel/next.js/blob/cd4798415259e932a18b91e883cb76f754067f0a/packages/next/build/index.ts#L389

**Reply by magne4000** [reply]
Yes it does catch `/assets` but not `/assets/*`. I also tried with a broader `^/(.*)` regex, same result

**Reply by barelyhuman** [reply]
I haven't really gone through the manifest handling completely, I'll go through that and ping in a few hours?

**Reply by magne4000** [reply]
Cool thanks, I'm trying to create a small reproducible case in the mean time

**Reply by magne4000** [reply]
It seems that when I put `source: '/'` it only matches `/`, and just ignores the regex (with this unique rule in the whole routes-manifest.json file)

**Reply by magne4000** [reply]
Tried the following with `dynamicRoutes` and it seems to work:
```json
{
  "version": 3,
  "basePath": "/",
  "pages404": true,
  "dynamicRoutes": [
    {
      "page": "/api/ssr_",
      "regex": "^/((?!assets/)(?!api/).*)$"
    }
  ]
}
```

This makes me think that there is some undocumented behavior around `rewrites[].source` behavior.

**Reply by barelyhuman** [reply]
The pattern system is a bit different for `vercel.json` and the `manifest.json`, one handles regex's one doesn't (not completely).

You can see the difference here 
https://vercel.com/docs/project-configuration#project-configuration/rewrites

Though, if the attempt you are making is just for the `routes-manifest.json` to work with vercel deployments then I guess it should follow the above standard but, if you are trying to get nextjs working then the regex's would actually work as intended

**Reply by barelyhuman** [reply]
I'm still going through the internal code for both the generation and parsing and trying out things but don't have a concrete explanation right now

**Comment by magne4000** [comment]
Related discussion https://github.com/vercel/vercel/discussions/7607
